java-checks/src/main/java/org/sonar/java/checks/RegexPatternsNeedlesslyCheck.java

```diff
@@ -0,0 +1,104 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.Arrays;
+import java.util.List;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.LiteralTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S4248")
+public class RegexPatternsNeedlesslyCheck extends AbstractMethodDetection {
+
+  private static final String STRING = "java.lang.String";
+  private static final String PATTERN = "java.util.regex.Pattern";
+  private static final char[] array;
+
+  static {
+    array = ".$|()[{^?*\\".toCharArray();
+    Arrays.sort(array);
+  }
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition(PATTERN).name("compile").addParameter(STRING),
+      MethodMatcher.create().typeDefinition(STRING).name("matches").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(STRING).name("split").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(STRING).name("replaceAll").withAnyParameters(),
+      MethodMatcher.create().typeDefinition(STRING).name("replaceFirst").withAnyParameters());
+  }
+
+  @Override
+  protected void onMethodInvocationFound(MethodInvocationTree mit) {
+    ExpressionTree argument = mit.arguments().get(0);
+    if (mit.symbolType().is(PATTERN)) {
+      if (argument.is(Tree.Kind.NULL_LITERAL) || (checkArgs(argument) && invokedOnlyOnce(mit))) {
+        return;
+      }
+    } else {
+      if (!argument.is(Tree.Kind.STRING_LITERAL) ||
+        ("split".equals(((MemberSelectExpressionTree) mit.methodSelect()).identifier().name()) && specialCaseNotReported(argument))) {
+        return;
+      }
+    }
+    reportIssue(mit, "Refactor this code to use a \"static final\" Pattern.");
+  }
+
+  private static boolean checkArgs(ExpressionTree argument) {
+    return argument.is(Tree.Kind.STRING_LITERAL) || (argument.is(Tree.Kind.IDENTIFIER) && (((IdentifierTree) argument).symbol().isFinal()));
+
+  }
+
+  private static boolean invokedOnlyOnce(MethodInvocationTree mit) {
+    Tree parentTree = mit.parent();
+    while (!parentTree.is(Tree.Kind.METHOD, Tree.Kind.CLASS)) {
+      if (parentTree.is(Tree.Kind.VARIABLE)) {
+        return ((VariableTree) parentTree).symbol().isFinal();
+      }
+      parentTree = parentTree.parent();
+    }
+    return false;
+  }
+
+  private static boolean specialCaseNotReported(ExpressionTree argument) {
```

Comment:
name is dodgy, I struggle to understand the code of this method and what it is supposed to do. The fact that this method should only apply to split might be mentioned in its name...